extension/httpfs/httpfs_curl_client.cpp

```diff
@@ -0,0 +1,486 @@
+// number of common locations and use the first one we find.
+static std::string certFileLocations[] = {
+	// Arch, Debian-based, Gentoo
+	"/etc/ssl/certs/ca-certificates.crt",
+	// RedHat 7 based
+	"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
+	// Redhat 6 based
+	"/etc/pki/tls/certs/ca-bundle.crt",
+	// OpenSUSE
+	"/etc/ssl/ca-bundle.pem",
+	// Alpine
+	"/etc/ssl/cert.pem"};
+
+
+//! Grab the first path that exists, from a list of well-known locations
+static std::string SelectCURLCertPath() {
+	for (std::string &caFile : certFileLocations) {
+		struct stat buf;
+		if (stat(caFile.c_str(), &buf) == 0) {
+			return caFile;
+		}
+	}
+	return std::string();
+}
+
+static std::string cert_path = SelectCURLCertPath();
+
+static size_t RequestWriteCallback(void *contents, size_t size, size_t nmemb, void *userp) {
+	size_t totalSize = size * nmemb;
+	std::string* str = static_cast<std::string*>(userp);
+	str->append(static_cast<char*>(contents), totalSize);
+	return totalSize;
+}
+
+static size_t RequestHeaderCallback(void *contents, size_t size, size_t nmemb, void *userp) {
+	size_t totalSize = size * nmemb;
+	std::string header(static_cast<char*>(contents), totalSize);
+	HeaderCollector* header_collection = static_cast<HeaderCollector*>(userp);
+
+	// Trim trailing \r\n
+	if (!header.empty() && header.back() == '\n') {
+		header.pop_back();
+		if (!header.empty() && header.back() == '\r') {
+			header.pop_back();
+		}
+	}
+
+	// If header starts with HTTP/... curl has followed a redirect and we have a new Header,
+	// so we push back a new header_collection and store headers from the redirect there.
+	if (header.rfind("HTTP/", 0) == 0) {
+		header_collection->header_collection.push_back(HTTPHeaders());
+		header_collection->header_collection.back().Insert("__RESPONSE_STATUS__", header);
+	}
+
+	size_t colonPos = header.find(':');
+
+	if (colonPos != std::string::npos) {
+		// Split the string into two parts
+		std::string part1 = header.substr(0, colonPos);
+		std::string part2 = header.substr(colonPos + 1);
+		if (part2.at(0) == ' ') {
+			part2.erase(0, 1);
+		}
+
+		header_collection->header_collection.back().Insert(part1, part2);
+	}
+	// TODO: log headers that don't follow the header format
+
+	return totalSize;
+}
+
+ CURLHandle::CURLHandle(const string &token, const string &cert_path) {
+	curl = curl_easy_init();
+	if (!curl) {
+		throw InternalException("Failed to initialize curl");
+	}
+	if (!token.empty()) {
+		curl_easy_setopt(curl, CURLOPT_XOAUTH2_BEARER, token.c_str());
+		curl_easy_setopt(curl, CURLOPT_HTTPAUTH, CURLAUTH_BEARER);
+	}
+	if (!cert_path.empty()) {
+		curl_easy_setopt(curl, CURLOPT_CAINFO, cert_path.c_str());
+	}
+}
+
+CURLHandle::~CURLHandle() {
+	curl_easy_cleanup(curl);
+}
+
+
+struct RequestInfo {
+	string url = "";
+	string body = "";
+	uint16_t response_code = 0;
+	std::vector<HTTPHeaders> header_collection;
+};
+
+
+static idx_t httpfs_client_count = 0;
```

Comment:
curious why don't we use `std::atomic<idx_t>` here? So we don't need another global variable (in two places...) and prob slightly better perf